Readme.md
![module info](https://nodei.co/npm/iofs.png?downloads=true&downloadRank=true&stars=true)

# iofs
> `iofs`是一个基于原生`fs`模块封装的工具, 旨在提供更加方便实用一些常用的API方法(同步), API习惯参考了`bash shell`, 习惯用命令行的朋友, 可能会比较亲切。


## API
+ props
  - [origin](#origin)
+ methods
  - [.cat(file)](#catfile)
  - [.ls(path, recursive)](#lspath-recursive)
  - [.echo(data, file, append, encode)](#echodata-file-append-encode)
  - [.chmod(path, mode)](#chmodpath-mode)
  - [.chown(path, uid, gid)](#chownpath-uid-gid)
  - [.mv(origin, target)](#mvorigin-target)
  - [.cp(origin, target)](#cporigin-target)
  - [.rm(origin)](#rmorigin)
  - [.stat(path)](#statpath)
  - [.isdir(path)](#isdirpath)
  - [.isfile(path)](#isfilepath)
  - [.mkdir(dir, mode)](#mkdirdir-mode)
  - [.exists(path)](#existspath)
  - [.is(path)](#isspath-mode)


## 属性 

### origin
> 返回原生的`fs`模块对象, 方便调用一些未封装的额外功能



## APIs
> 所有API均支持在最后传入一个 `debug<Boolean>`参数(v1.3.2新增), 用于打印错误日志

### .cat(file)
> 读取文件, 返回一个`Buffer对象`

| 参数 | 类型 | 是否必须 | 说明 |
| :--: | :--: | :--: | -- |
| file |  `<String>`  |  是  | 要读取的文件路径 |

---


### .ls(path, recursive)
> 列出指定目录下的所有文件&目录, 不包括 '.' and '..'. 结果返回一个数组.

| 参数 | 类型 | 是否必须 | 说明 |
| :--: | :--: | :--: | -- |
| path |  `<String>`  |  是  | 要读取的目录 |
| recursive |  `<String>`  |  否  | 是否递归读取 |

---

### .echo(data, file, append, encode)
> 写数据到指定文件中. 如果指定文件不存在, 则自动生成.

| 参数 | 类型 | 是否必须 | 说明 |
| :--: | :--: | :--: | -- |
| data |  `<String>` `<Buffer>` `<Number>` |  是  | 要写入的数据, 可以字符串、Buffer对象, 数字 |
| file |  `<String>`  |  是  | 要写入的文件名, 不存在会自动创建, 如存在会覆盖 |
| append |  `<Boolean>`  |  否  | 是否在文件后追加数据, 默认否, 即会整个文件替换 |
| encode |  `<String>`  |  否  | 指定保存的编码, 默认utf8 |


```javascript
var fs = require('iofs')

fs.echo('hello ', 'test.txt') // 如果test.txt存在, 则覆盖.
fs.echo('world', 'test.txt', true) // 不会覆盖, 只会追加到 test.txt中

```

---


### .chmod(path, mode)
> 修改文件&目录的权限.

| 参数 | 类型 | 是否必须 | 说明 |
| :--: | :--: | :--: | -- |
| path |  `<String>`|  是  | 要修改的文件&目录路径 |
| mode |  `<Number>`  |  是  | 权限码 `0o000 - 0o777` |


```javascript

fs.chmod('test.txt', 0o777)

```

---


### .chown(path, uid, gid)
> 修改文件&目录的归属。
>> `v1.3.0 新增`

| 参数 | 类型 | 是否必须 | 说明 |
| :--: | :--: | :--: | -- |
| path |  `<String>`|  是  | 要修改的文件&目录路径 |
| uid |  `<Number>`  |  是  | 用户ID |
| gid |  `<Number>`  |  是  | 用户组ID |



---




### .mv(origin, target)
> 移动文件&目录, 支持跨磁盘移动; 同时具备重命名功能。
>> `v1.3.0 之后支持对目录进行操作`

| 参数 | 类型 | 是否必须 | 说明 |
| :--: | :--: | :--: | -- |
| origin |  `<String>`|  是  | 要移动或重命名的文件&目录 |
| target |  `<String>`  |  是  | 目标文件名&目录名 |


---


### .cp(origin, target)
> 复制文件&目录, 支持跨磁盘复制。
>> `v1.3.0 之后支持对目录进行操作`

| 参数 | 类型 | 是否必须 | 说明 |
| :--: | :--: | :--: | -- |
| origin |  `<String>`|  是  | 要复制的文件&目录 |
| target |  `<String>`  |  是  | 目标文件名&目录名 |


---


### .rm(origin)
> 删除文件&目录
>> `v1.3.0 之后取消第2个参数, 改为自动判断是否目录, 是否自动递归删除`

| 参数 | 类型 | 是否必须 | 说明 |
| :--: | :--: | :--: | -- |
| origin |  `<String>`|  是  | 要删除的文件&目录 |

```javascript

fs.rm('./foo/test.txt')
fs.rm('./foo') // 整个目录删除

```


---



### .stat(path)
> 返回文件&目录的状态信息, 如修改时间, 文件大小等

| 参数 | 类型 | 是否必须 | 说明 |
| :--: | :--: | :--: | -- |
| path |  `<String>`|  是  | 要读取的目录&文件 |




---


### .isdir(path)
> 判断指定目录是否为一个目录, 路径不存在或者不是目录都会返回 false.

| 参数 | 类型 | 是否必须 | 说明 |
| :--: | :--: | :--: | -- |
| path |  `<String>`|  是  | 要读取的目录路径 |

---

### .isfile(path)
> 判断指定目录是否为一个文件, 路径不存在或者不是文件都会返回 false

| 参数 | 类型 | 是否必须 | 说明 |
| :--: | :--: | :--: | -- |
| path |  `<String>`|  是  | 要读取的文件 |

---

### .mkdir(dir)
> 创建目录, 会自动创建上级目录(如不存在)

| 参数 | 类型 | 是否必须 | 说明 |
| :--: | :--: | :--: | -- |
| dir |  `<String>`|  是  | 要创建的目录名 |


---

### .exists(path)
> 判断文件&目录是否存在

| 参数 | 类型 | 是否必须 | 说明 |
| :--: | :--: | :--: | -- |
| path |  `<String>`|  是  | 要读取的目录&文件 |


---

### .is(path, mode)
> 判断文件&目录是否存在

| 参数 | 类型 | 是否必须 | 说明 |
| :--: | :--: | :--: | -- |
| path |  `<String>`|  是  | 要读取的目录&文件 |
| mode |  `<Number>`|  否  | 如 r: 4, w: 2, rw: 6 |




History.md
# 1.4.0 / 2020-09-19
* [A] 增加esm模式

# 1.3.2 / 2020-01-18
* [U] 增加静默参数, 可以不打印错误日志(用于某些特殊场景)


# 1.3.1 / 2020-01-09
* [U] 优化异常输出 


# 1.3.0 / 2020-01-08
* [A] 增加chown()方法 
* [A] cp()和 mv()支持对目录进行操作
* [U] 优化异常处理, 友好的输出, 不终止代码往下执行
* [U] 支持新版本node使用新API
* [U] rm()方法自动识别目录,不再需要提供第2个参数


# 1.2.2 / 2020-01-02
* [F] 修复改回同步时回调未移除, 导致在node v12.10报错的bug 


# 1.2.1 / 2019-08-06
* [U] 写操作改回同步

# 1.2.0 / 2019-07-16
* [U] 优化结构
* [U] 部分写操作改为异步


# 1.1.0 / 2018-05-24
* [A] 增加异常信息输出
* [U] 优化`ls`API
* [U] 重命名 `this.self` 为 `this.origin`



# 1.0.3 / 2017-12-23
* [D] 删除异常信息输出



# 1.0.2 / 2017-11-20
* [F] 修复非空目录无法删除的bug



# 1.0.1 / 2017-04-24
* [A] `ls`方法当第2个参数为`true`的时候,可以遍历子目录


# 1.0.0 / 2017-02-26
* 初始化项目


index.mjs
/**
 * @author yutent<yutent.io@gmail.com>
 * @date 2020/09/19 16:39:59
 */

import FS from 'fs'
import PATH from 'path'

class STATS {
  isFile() {
    return false
  }
  isDirectory() {
    return false
  }
  isSocket() {
    return false
  }
  isSymbolicLink() {
    return false
  }
}

const VERSION = +process.versions.node
  .split('.')
  .slice(0, 2)
  .join('.')

const EMPTY_STAT = new STATS()

export default {
  origin: FS,

  /**
   * [cat 文件读取]
   * @param  {String} file [文件路径]
   * @param  {Function} cb   [回调] 可选
   */
  cat(file, debug) {
    try {
      return FS.readFileSync(file)
    } catch (err) {
      debug && console.error('call cat(): ', err + '')
      return null
    }
  },

  /**
   * [ls 读取整个目录(不遍历子目录)]
   * @param  {string} dir [目标路径]
   * @param  {boolean} recursive [是否递归遍历子目录]
   * @return {array}      [返回目标目录所有文件名和子目录名, 不包括'.'和'..']
   */
  ls(dir, recursive, debug) {
    try {
      var list = FS.readdirSync(dir)

      list.forEach((it, i) => {
        list[i] = PATH.resolve(dir, it)
      })

      if (recursive) {
        var tmp = list.concat()
        tmp.forEach(it => {
          if (this.isdir(it)) {
            list = list.concat(this.ls(it))
          }
        })
      }
      return list
    } catch (err) {
      debug && console.error('call ls(): ', err + '')
      return null
    }
  },

  /**
   * [echo 写文件]
   * @param  {String|Buffer|Number} data   [要写入的数据]
   * @param  {String} file   [要写的文件]
   * @param  {Boolean} append [是否在后面追加，默认否]
   * @param  {String} encode [编码, 默认utf8]
   */
  echo(data, file, append, encode, debug) {
    if (!file) {
      return data
    }

    var updir = PATH.parse(file).dir
    var opt = {}
    if (!this.isdir(updir)) {
      this.mkdir(updir)
    }

    if (append && typeof append === 'string') {
      encode = append
      append = false
      opt.encoding = encode
    } else {
      if (typeof encode === 'string') {
        opt.encoding = encode
      }
    }

    try {
      if (!!append) {
        FS.appendFileSync(file, data, opt)
      } else {
        FS.writeFileSync(file, data, opt)
      }
      return true
    } catch (err) {
      debug && console.error('call echo(): ', err + '')
      return false
    }
  },

  //修改权限
  chmod(path, mode, debug) {
    try {
      FS.chmodSync(path, mode)
      return true
    } catch (err) {
      debug && console.error('call chmod(): ', err + '')
      return false
    }
  },

  //修改所属用户
  chown(path, uid, gid, debug) {
    try {
      FS.chownSync(path, uid, gid)
      return true
    } catch (err) {
      debug && console.error('call chown(): ', err + '')
      return false
    }
  },

  /**
   * [mv 移动文件&目录,兼具重命名功能]
   * @param  {String} origin [原路径/原名]
   * @param  {String} target   [目标路径/新名]
   */
  mv(origin, target, debug) {
    var updir = PATH.parse(target).dir
    if (!this.isdir(updir)) {
      this.mkdir(updir)
    }

    try {
      FS.renameSync(origin, target)
    } catch (err) {
      if (~err.message.indexOf('cross-device')) {
        if (this.cp(origin, target)) {
          return this.rm(origin)
        }
        return false
      }
      debug && console.error('call mv(): ', err + '')
      return false
    }
  },

  /**
   * [cp 复制文件&目录]
   * @param  {String} origin [原路径]
   * @param  {String} target   [目标路径]
   */
  cp(origin, target, debug) {
    try {
      // 如果是目录, 则递归操作
      if (this.isdir(origin)) {
        this.mkdir(target)
        var list = this.ls(origin)
        list.forEach(val => {
          let name = PATH.parse(val).base
          this.cp(val, PATH.join(target, name))
        })
      } else {
        var updir = PATH.parse(target).dir
        if (!this.isdir(updir)) {
          this.mkdir(updir)
        }

        var rs = FS.createReadStream(origin)
        var ws = FS.createWriteStream(target)
        rs.pipe(ws)
      }

      return true
    } catch (err) {
      debug && console.error('call cp(): ', err + '')
      return false
    }
  },

  /**
   * [rm 删除文件/目录]
   * @param  {[type]} origin      [源文件/目录路径]
   */
  rm(origin, debug) {
    try {
      if (this.isdir(origin)) {
        if (VERSION > 12.1) {
          FS.rmdirSync(origin, { recursive: true })
        } else {
          var list = this.ls(origin)
          list.forEach(it => this.rm(it))
          FS.rmdirSync(origin)
        }
      } else {
        FS.unlinkSync(origin)
      }
      return true
    } catch (err) {
      debug && console.error('call rm(): ', err + '')
      return false
    }
  },

  /**
   * [stat 返回文件/目录的状态信息]
   * @param  {[string]} path [目标路径]
   * @param  {[boolean]} debug [是否静默检测, 是否不打印错误日志]
   */
  stat(path, debug) {
    try {
      return FS.statSync(path)
    } catch (err) {
      debug && console.error('call stat(): ', err + '')
      return EMPTY_STAT
    }
  },

  /**
   * [isdir 判断目标是否为目录]
   * @param  {String} path [目标路径]
   */
  isdir(path) {
    try {
      return this.stat(path).isDirectory()
    } catch (err) {
      return false
    }
  },

  isfile(path) {
    try {
      return this.stat(path).isFile()
    } catch (err) {
      return false
    }
  },

  /**
   * [mkdir 新建目录]
   * @param  {String} dir [目标路径]
   * @param {Number} mode [目录权限, node v10.12起支持]
   */
  mkdir(dir, mode = 0o755, debug) {
    try {
      if (VERSION > 10.12) {
        FS.mkdirSync(dir, { recursive: true, mode: mode })
      } else {
        var updir = PATH.parse(dir).dir
        if (!updir) {
          debug && console.error('call mkdir(): ', 'Wrong dir path')
          return false
        }

        if (!this.isdir(updir)) {
          this.mkdir(updir)
        }

        FS.mkdirSync(dir)
        this.chmod(dir, mode)
      }
      return true
    } catch (err) {
      debug && console.error('call mkdir(): ', err + '')
      return false
    }
  },

  /**
   * [exists 判断目标(文件/目录)是否存在]
   * @param  {String} file [目标路径]
   */
  exists(file) {
    return this.is(file, FS.constants.R_OK)
  },

  // 是否可读写
  is(file, mode) {
    try {
      FS.accessSync(file, mode)
      return true
    } catch (e) {
      return false
    }
  }
}


index.js
/**
 * @author yutent<yutent.io@gmail.com>
 * @date 2020/09/19 16:39:59
 */

const FS = require('fs')
const PATH = require('path')

class Stats {
  isFile() {
    return false
  }
  isDirectory() {
    return false
  }
  isSocket() {
    return false
  }
  isSymbolicLink() {
    return false
  }
}

const VERSION = +process.versions.node
  .split('.')
  .slice(0, 2)
  .join('.')

const EMPTY_STAT = new Stats()

const Iofs = {
  origin: FS,

  /**
   * [cat 文件读取]
   * @param  {String} file [文件路径]
   * @param  {Function} cb   [回调] 可选
   */
  cat(file, debug) {
    try {
      return FS.readFileSync(file)
    } catch (err) {
      debug && console.error('call cat(): ', err + '')
      return null
    }
  },

  /**
   * [ls 读取整个目录(不遍历子目录)]
   * @param  {string} dir [目标路径]
   * @param  {boolean} recursive [是否递归遍历子目录]
   * @return {array}      [返回目标目录所有文件名和子目录名, 不包括'.'和'..']
   */
  ls(dir, recursive, debug) {
    try {
      var list = FS.readdirSync(dir)

      list.forEach((it, i) => {
        list[i] = PATH.resolve(dir, it)
      })

      if (recursive) {
        var tmp = list.concat()
        tmp.forEach(it => {
          if (this.isdir(it)) {
            list = list.concat(this.ls(it))
          }
        })
      }
      return list
    } catch (err) {
      debug && console.error('call ls(): ', err + '')
      return null
    }
  },

  /**
   * [echo 写文件]
   * @param  {String|Buffer|Number} data   [要写入的数据]
   * @param  {String} file   [要写的文件]
   * @param  {Boolean} append [是否在后面追加，默认否]
   * @param  {String} encode [编码, 默认utf8]
   */
  echo(data, file, append, encode, debug) {
    if (!file) {
      return data
    }

    var updir = PATH.parse(file).dir
    var opt = {}
    if (!this.isdir(updir)) {
      this.mkdir(updir)
    }

    if (append && typeof append === 'string') {
      encode = append
      append = false
      opt.encoding = encode
    } else {
      if (typeof encode === 'string') {
        opt.encoding = encode
      }
    }

    try {
      if (!!append) {
        FS.appendFileSync(file, data, opt)
      } else {
        FS.writeFileSync(file, data, opt)
      }
      return true
    } catch (err) {
      debug && console.error('call echo(): ', err + '')
      return false
    }
  },

  //修改权限
  chmod(path, mode, debug) {
    try {
      FS.chmodSync(path, mode)
      return true
    } catch (err) {
      debug && console.error('call chmod(): ', err + '')
      return false
    }
  },

  //修改所属用户
  chown(path, uid, gid, debug) {
    try {
      FS.chownSync(path, uid, gid)
      return true
    } catch (err) {
      debug && console.error('call chown(): ', err + '')
      return false
    }
  },

  /**
   * [mv 移动文件&目录,兼具重命名功能]
   * @param  {String} origin [原路径/原名]
   * @param  {String} target   [目标路径/新名]
   */
  mv(origin, target, debug) {
    var updir = PATH.parse(target).dir
    if (!this.isdir(updir)) {
      this.mkdir(updir)
    }

    try {
      FS.renameSync(origin, target)
    } catch (err) {
      if (~err.message.indexOf('cross-device')) {
        if (this.cp(origin, target)) {
          return this.rm(origin)
        }
        return false
      }
      debug && console.error('call mv(): ', err + '')
      return false
    }
  },

  /**
   * [cp 复制文件&目录]
   * @param  {String} origin [原路径]
   * @param  {String} target   [目标路径]
   */
  cp(origin, target, debug) {
    try {
      // 如果是目录, 则递归操作
      if (this.isdir(origin)) {
        this.mkdir(target)
        var list = this.ls(origin)
        list.forEach(val => {
          let name = PATH.parse(val).base
          this.cp(val, PATH.join(target, name))
        })
      } else {
        var updir = PATH.parse(target).dir
        if (!this.isdir(updir)) {
          this.mkdir(updir)
        }

        var rs = FS.createReadStream(origin)
        var ws = FS.createWriteStream(target)
        rs.pipe(ws)
      }

      return true
    } catch (err) {
      debug && console.error('call cp(): ', err + '')
      return false
    }
  },

  /**
   * [rm 删除文件/目录]
   * @param  {[type]} origin      [源文件/目录路径]
   */
  rm(origin, debug) {
    try {
      if (this.isdir(origin)) {
        if (VERSION > 12.1) {
          FS.rmdirSync(origin, { recursive: true })
        } else {
          var list = this.ls(origin)
          list.forEach(it => this.rm(it))
          FS.rmdirSync(origin)
        }
      } else {
        FS.unlinkSync(origin)
      }
      return true
    } catch (err) {
      debug && console.error('call rm(): ', err + '')
      return false
    }
  },

  /**
   * [stat 返回文件/目录的状态信息]
   * @param  {[string]} path [目标路径]
   * @param  {[boolean]} debug [是否静默检测, 是否不打印错误日志]
   */
  stat(path, debug) {
    try {
      return FS.statSync(path)
    } catch (err) {
      debug && console.error('call stat(): ', err + '')
      return EMPTY_STAT
    }
  },

  /**
   * [isdir 判断目标是否为目录]
   * @param  {String} path [目标路径]
   */
  isdir(path) {
    try {
      return this.stat(path).isDirectory()
    } catch (err) {
      return false
    }
  },

  isfile(path) {
    try {
      return this.stat(path).isFile()
    } catch (err) {
      return false
    }
  },

  /**
   * [mkdir 新建目录]
   * @param  {String} dir [目标路径]
   * @param {Number} mode [目录权限, node v10.12起支持]
   */
  mkdir(dir, mode = 0o755, debug) {
    try {
      if (VERSION > 10.12) {
        FS.mkdirSync(dir, { recursive: true, mode: mode })
      } else {
        var updir = PATH.parse(dir).dir
        if (!updir) {
          debug && console.error('call mkdir(): ', 'Wrong dir path')
          return false
        }

        if (!this.isdir(updir)) {
          this.mkdir(updir)
        }

        FS.mkdirSync(dir)
        this.chmod(dir, mode)
      }
      return true
    } catch (err) {
      debug && console.error('call mkdir(): ', err + '')
      return false
    }
  },

  /**
   * [exists 判断目标(文件/目录)是否存在]
   * @param  {String} file [目标路径]
   */
  exists(file) {
    return this.is(file, FS.constants.R_OK)
  },

  // 是否可读写
  is(file, mode) {
    try {
      FS.accessSync(file, mode)
      return true
    } catch (e) {
      return false
    }
  }
}

module.exports = Iofs
